### ISSUE TITLE
MultiAgentEnvironment: support multi-discrete action spaces

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#3]
## Motivation

\`MultiAgentEnvironment::step_multi(&[i64])\` assumes each agent emits a **single** discrete action. Many real environments have **factored / multi-discrete** action spaces:

- **Bucket Brigade**: \`[house_index, mode]\` — \`[10, 2]\`
- **StarCraft micro**: \`[action_type, target_unit, position]\`
- **Card / board games**: \`[card_to_play, target, modifier]\`

Today these have to be packed into a single i64 (cardinality = product), which:
1. Inflates the policy head's output dim multiplicatively (\`10 * 2 = 20\` for BB; explodes for SC2).
2. Loses the factored independence structure PPO can exploit (separate categoricals per dim → summed log-probs).
3. Forces the env to expose a custom encode/decode helper.

## Proposal

Widen the trait method to accept a flat slice per agent, paired with a per-agent shape descriptor:

\`\`\`rust
pub trait MultiAgentEnvironment: Environment {
    fn num_agents(&self) -> usize;
    fn agent_action_space(&self, agent_id: usize) -> Vec<usize>; // e.g. vec![10, 2]
    fn step_multi(&mut self, actions: &[Vec<i64>]) -> MultiAgentResult;
    ...
}
\`\`\`

Alternative: a new \`MultiDiscreteMultiAgentEnvironment\` trait so the existing single-discrete path stays simple.

The \`MlpPolicy\` will also need a multi-head variant — \`Vec<nn::Linear>\` for action heads — and a matching loss path (sum of per-dim log-probs / mean of per-dim entropies). This is the bulk of the actual work; the trait change is a 30-line patch.

## Use case

Slepian-Wolf MARL P3 experiment on Bucket Brigade (\`bucket-brigade-core\`) needs \`[house, mode]\` factored actions. Our current Python prototype (\`bucket_brigade.training.joint_trainer.JointPPOTrainer\`) implements the factored path; porting to thrust requires this.

## Related

Pre-requisite for #6 (BB env adapter).

**Comment by rjwalters** [comment]
## Curator Enhancement

Curating per user request. Verified against `master` as of commit `1ddbbf5` (Rename default branch master -> main) -- the `slepian-wolf-marl-port` branch has been merged, but this issue remains open work: the `MultiAgentEnvironment` trait still takes `&[i64]`, and `BucketBrigadeMaEnv` deliberately bypasses the trait.

### Codebase Verification (May 2026)

**Confirmed: this work is still open.**

- The trait method still has the narrow signature:
  ```rust
  // src/multi_agent/environment.rs:36
  fn step_multi(&mut self, actions: &[i64]) -> MultiAgentResult;
  ```
- The new factored policy (`MultiDiscreteMlpPolicy`) landed in commit `c5030a2` and lives at `src/policy/multi_discrete_mlp.rs`. It is fully functional with per-dim logits/log-probs/entropy and is already used by `examples/games/bucket_brigade/train_p3.rs`.
- The new `BucketBrigadeMaEnv` (commit `273bbd5`, `src/env/games/bucket_brigade.rs`) explicitly documents the gap and does *not* implement the trait:
  ```rust
  // src/env/games/bucket_brigade.rs:17-20
  //! Does **not** implement [`crate::multi_agent::MultiAgentEnvironment`]
  //! because that trait currently takes `&[i64]` and doesn't support factored
  //! action spaces (see upstream issue #3). When the trait is widened, this
  //! struct can grow a trivial impl.
  ```
  Its inherent `step()` takes `&[[u8; 2]]` (i.e. per-agent `[house_index, mode]`).

**So the policy half is already done.** This issue is now narrowly scoped to widening the trait + simulator + the lone real impl, and adding the `BucketBrigadeMaEnv` impl as the validation case (cleanly resolving the FIXME comment quoted above).

### Root Cause / Why This Matters

`MultiAgentEnvironment::step_multi(&[i64])` forces a single discrete action per agent. Factored action spaces (Bucket Brigade `[house, mode]`, StarCraft micro `[type, target, position]`, card games `[card, target, modifier]`) must be packed into a Cartesian-product i64:

1. **Policy head explosion**: `n_houses * n_modes = 10 * 2 = 20` for BB; combinatorial for SC2.
2. **Loses factored independence**: PPO can't exploit separable per-dim categoricals (the implementation in `MultiDiscreteMlpPolicy::evaluate_actions` sums per-dim log-probs and averages per-dim entropies -- that whole channel is wasted if the env collapses everything to one i64).
3. **Forces env-specific encode/decode shims**: `BucketBrigadeMaEnv` already exposes `action_dims() -> [i64; 2]` precisely because it can't speak the trait's narrow vocabulary.

### Implementation Guidance

The proposal in the original issue is sound; here is the concrete shape.

#### Approach A (recommended): widen the existing trait

Add a per-agent action-space descriptor and replace the scalar slice with a slice of `Vec<i64>` (one inner Vec per agent, length = number of action dims for that agent).

```rust
// src/multi_agent/environment.rs
pub trait MultiAgentEnvironment: Environment {
    fn num_agents(&self) -> usize;
    fn get_agent_observation(&self, agent_id: usize) -> Vec<f32>;

    /// Per-agent action-space layout. For a pure-discrete agent: `vec![n]`.
    /// For factored Bucket Brigade: `vec![10, 2]`. Length must match
    /// the policy's `action_dims()` for that agent.
    fn agent_action_space(&self, agent_id: usize) -> Vec<usize>;

    fn step_multi(&mut self, actions: &[Vec<i64>]) -> MultiAgentResult;
    fn active_agents(&self) -> Vec<bool>;
}
```

**Backward-compat consideration**: every existing impl uses scalar actions today. Pure-discrete envs would wrap each scalar as a length-1 `Vec<i64>`. This is intrusive but localized -- the only *real* impl is the mock in `simulator.rs` tests (`MockEnv`); the snake `MultiAgentSnakeEnv` impl is **commented out** in `src/env/games/snake.rs:23,31` and not currently wired in, so it doesn't constrain us.

#### Approach B: new sibling trait

Introduce `MultiDiscreteMultiAgentEnvironment` alongside the existing trait:

```rust
pub trait MultiDiscreteMultiAgentEnvironment: Environment {
    fn num_agents(&self) -> usize;
    fn get_agent_observation(&self, agent_id: usize) -> Vec<f32>;
    fn agent_action_space(&self, agent_id: usize) -> Vec<usize>;
    fn step_multi(&mut self, actions: &[Vec<i64>]) -> MultiAgentResult;
    fn active_agents(&self) -> Vec<bool>;
}
```

Pro: zero churn on existing single-discrete code path. Con: `GameSimulator` would need to be generic over both traits (or duplicated), and we accumulate parallel infrastructure.

**Recommendation: Approach A.** The blast radius is small (one real impl: the mock; one wasm consumer; one simulator) and the long-term simpler trait is worth a small migration. Pure-discrete callers just wrap actions in length-1 vecs, which is a 1-line site change.

#### Sub-tasks

1. **Trait widening** (`src/multi_agent/environment.rs`):
   - Change `step_multi` signature to `&[Vec<i64>]`.
   - Add `agent_action_space(&self, agent_id: usize) -> Vec<usize>`.
   - The mock test in `src/multi_agent/environment.rs:84-189` is already commented out; leave it alone or rewrite it to the new shape.

2. **Simulator update** (`src/multi_agent/simulator.rs:131-207`):
   - Around line 144-167 the simulator collects `actions: Vec<i64>` from each policy via `policy.get_action(&obs_tensor)`. Today this returns a scalar; for multi-discrete it returns `[batch, num_dims]`. The simulator needs to gather per-agent action vectors instead of scalars and call `env.step_multi(&actions)` where `actions: Vec<Vec<i64>>`.
   - The `Population` / `Agent` policy abstraction currently assumes scalar actions (see `Experience` struct fields at `src/multi_agent/messages.rs` and the `action_val: i64 = action.int64_value(&[])` extraction at `src/multi_agent/simulator.rs:160`). The cleanest path is to widen `Experience::action` from `i64` to `Vec<i64>` (or `Tensor`), or introduce a parallel `MultiDiscreteExperience` variant. **Out of scope for this issue if it cascades** -- see scope guard below.

3. **BucketBrigadeMaEnv impl** (`src/env/games/bucket_brigade.rs`):
   - Add `impl MultiAgentEnvironment for BucketBrigadeMaEnv`. Translate `&[Vec<i64>]` to the inner `&[[u8; 2]]` shape:
     ```rust
     fn step_multi(&mut self, actions: &[Vec<i64>]) -> MultiAgentResult {
         let inner: Vec<[u8; 2]> = actions.iter()
             .map(|a| [a[0] as u8, a[1] as u8])
             .collect();
         let r = self.step(&inner);
         MultiAgentResult::new(r.observations, r.rewards,
             vec![r.done; self.num_agents],
             vec![false; self.num_agents])
     }
     fn agent_action_space(&self, _: usize) -> Vec<usize> { vec![10, 2] }
     ```
   - Resolves the FIXME at `src/env/games/bucket_brigade.rs:17-20`.

4. **WASM site** (`src/wasm.rs:170-176`):
   - Currently calls `self.env.step_multi(&actions_i64)` against `SnakeEnv::step_multi`, which is an **inherent method** on `SnakeEnv` taking `&[i64]` -- NOT the trait method. This call site is unaffected by the trait change. Leave it alone but verify with `cargo build --features wasm`.

5. **Snake call sites** (`examples/games/snake/train_snake_multi.rs:246`, `optimize_snake.rs:252`):
   - Same story: these call the inherent `SnakeEnv::step_multi(&[i64]) -> StepResult`, not the trait. Unaffected.

6. **Tests**:
   - Replace/update `src/multi_agent/simulator.rs:296-317` `MockEnv` impl to the new trait shape.
   - Add a new integration test: `tests/test_multi_discrete_env.rs` exercising `BucketBrigadeMaEnv` through the trait surface.

### Scope Guard

**In scope:**
- Trait signature change in `src/multi_agent/environment.rs`.
- `BucketBrigadeMaEnv` impl + removing the FIXME.
- Simulator updates necessary to call the new `step_multi` (signature change only -- not a deep refactor of the experience-collection path).
- Mock test update.

**Out of scope (file follow-ups if needed):**
- Widening `Experience`/`Population`/`PolicyLearner` to support multi-discrete actions end-to-end. If the simulator's `Experience::action: i64` field blocks the trait change from compiling, the implementer should either (a) hold the simulator's call site at a TODO with `unimplemented!()` for multi-discrete and continue accepting scalar actions for now (preserving GameSimulator for single-discrete only), or (b) open a follow-up issue. **The Bucket Brigade trainer in `examples/games/bucket_brigade/train_p3.rs` doesn't use `GameSimulator` at all** -- it has its own joint rollout (see `collect_rollout` at line 129 of that file) -- so the simulator path isn't a blocker for the BB experiment.
- Migrating snake to the multi-discrete trait (snake is still single-discrete; its `MultiAgentEnvironment` impl is commented out and a separate concern).

**Hard limits:**
- LOC budget: ~250 LOC across implementation + tests. If it blows past 400, decompose into trait change + BB impl + simulator follow-up.
- Files touched: ~5-6 (`environment.rs`, `bucket_brigade.rs`, `simulator.rs` mock, optionally `messages.rs` if Experience must widen, plus one new test file).

### Affected Files

- `src/multi_agent/environment.rs` -- widen `step_multi` to `&[Vec<i64>]`; add `agent_action_space`.
- `src/multi_agent/simulator.rs` -- update mock `MockEnv` impl (lines 296-317); update `run_episodes` action-collection path (lines 131-207) to produce `Vec<Vec<i64>>` instead of `Vec<i64>`. If the `Experience::action: i64` schema blocks the change, see Scope Guard above.
- `src/multi_agent/messages.rs` -- may need to widen `Experience::action` to `Vec<i64>` or `Tensor`; out-of-scope follow-up acceptable.
- `src/env/games/bucket_brigade.rs` -- add `impl MultiAgentEnvironment`; remove FIXME at lines 17-20.
- `tests/test_multi_discrete_env.rs` (new) -- integration test for the BB env through the new trait surface.
- `src/wasm.rs:170-176` -- verify build still passes (`SnakeEnv::step_multi` is inherent, not trait; should be unaffected).
- `examples/games/snake/*.rs` -- verify builds (same reasoning -- inherent method, not trait).

### Test Plan

- [ ] Manual verification: `cargo build --release --features training` and `cargo build --features wasm` both succeed.
- [ ] Manual verification: `cargo run --release --example train_p3 --features \"training env-bucket-brigade\" -- --num-iterations 2 --rollout-steps 64` runs to completion using the BB env through *both* its inherent API and (in a new path) the trait, producing identical trajectories.
- [ ] Automated tests: new `tests/test_multi_discrete_env.rs`:
  - [ ] `agent_action_space` returns `vec![10, 2]` for every BB agent.
  - [ ] `step_multi(&[vec![3, 1], vec![5, 0], ...])` produces a `MultiAgentResult` whose `rewards.len() == num_agents` and whose `observations[i].len() == obs_dim()`.
  - [ ] Calling `step_multi` with an action vec of the wrong length panics with a clear error.
- [ ] Automated tests: existing tests in `src/multi_agent/simulator.rs` (`test_simulator_creation`) still pass after the `MockEnv` mock is updated to the new trait signature.
- [ ] Edge cases: single-dim discrete agent (e.g. `vec![4]` for snake-style) still round-trips correctly through the new trait. Confirms the widening is a strict superset of the old behavior.
- [ ] Edge cases: empty action vec / mismatched `num_agents` length triggers an `assert!`/`panic!` with a helpful message.

### Acceptance Criteria

- [ ] `MultiAgentEnvironment::step_multi` accepts `&[Vec<i64>]` (one action vector per agent).
- [ ] `MultiAgentEnvironment::agent_action_space(agent_id) -> Vec<usize>` is defined and documents the per-agent dim layout.
- [ ] `BucketBrigadeMaEnv` implements `MultiAgentEnvironment` and the FIXME comment at `src/env/games/bucket_brigade.rs:17-20` is removed.
- [ ] The library compiles with `--features training`, `--features wasm`, and `--features env-bucket-brigade`.
- [ ] All existing tests pass; new BB trait-level test passes.
- [ ] `cargo clippy --all-targets --all-features -- -D warnings` is clean.
- [ ] If simulator-level multi-discrete support is deferred, a follow-up issue is filed explicitly tracking it.

### Related

- Pre-requisite for **#6** (BB env adapter -- `loom:curating` at the time of this enhancement; the two should be co-curated).
- Sibling to **#5** (synchronized `JointMultiAgentTrainer` -- the train_p3 example is the prototype; cleanly slots in once both #3 and #6 land).
- Consumes **#1** (`MlpPolicy::encoder_features`, merged) -- the cross-agent redundancy regularizer in train_p3 already uses the tap.
